File: Java/src/main/java/edu/rice/dmodel/Order.java
package edu.rice.dmodel;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import javax.json.Json;
import javax.json.JsonArray;
import javax.json.JsonArrayBuilder;
import javax.json.JsonObject;
import javax.json.JsonObjectBuilder;
import javax.json.JsonReader;
import javax.json.JsonWriter;
import javax.json.JsonWriterFactory;
import javax.json.stream.JsonGenerator;

public class Order extends Base implements RootData {

	// CREATE TABLE ORDERS ( O_ORDERKEY INTEGER NOT NULL,
	// O_CUSTKEY INTEGER NOT NULL,
	// O_ORDERSTATUS CHAR(1) NOT NULL,
	// O_TOTALPRICE DECIMAL(15,2) NOT NULL,
	// O_ORDERDATE DATE NOT NULL,
	// O_ORDERPRIORITY CHAR(15) NOT NULL,
	// O_CLERK CHAR(15) NOT NULL,
	// O_SHIPPRIORITY INTEGER NOT NULL,
	// O_COMMENT VARCHAR(79) NOT NULL);

	// private int orderkey;
	// private int custkey;

	private static final long serialVersionUID = 6828883074098155463L;

	private List<LineItem> lineItems;

	private int orderkey;
	private int custkey;
	private String orderstatus;
	private double totalprice;
	private String orderdate;
	private String orderpriority;
	private String clerk;
	private int shippriority;
	private String comment;

	public Order() {

	}

	public Order(List<LineItem> lineItems, int orderkey, int custkey, String orderstatus, double totalprice, String orderdate, String orderpriority, String clerk, int shippriority, String comment) {
		super();

		this.lineItems = lineItems;
		this.orderkey = orderkey;
		this.custkey = custkey;
		this.orderstatus = orderstatus;
		this.totalprice = totalprice;
		this.orderdate = orderdate;
		this.orderpriority = orderpriority;
		this.clerk = clerk;
		this.shippriority = shippriority;
		this.comment = comment;
	}

	public int getOrderkey() {
		return orderkey;
	}

	public void setOrderkey(int orderkey) {
		this.orderkey = orderkey;
	}

	public int getCustkey() {
		return custkey;
	}

	public void setCustkey(int custkey) {
		this.custkey = custkey;
	}

	public double getTotalprice() {
		return totalprice;
	}

	public void setTotalprice(double totalprice) {
		this.totalprice = totalprice;
	}

	public String getOrderstatus() {
		return orderstatus;
	}

	public void setOrderstatus(String orderstatus) {
		this.orderstatus = orderstatus;
	}

	public String getOrderdate() {
		return orderdate;
	}

	public void setOrderdate(String orderdate) {
		this.orderdate = orderdate;
	}

	public String getOrderpriority() {
		return orderpriority;
	}

	public void setOrderpriority(String orderpriority) {
		this.orderpriority = orderpriority;
	}

	public String getClerk() {
		return clerk;
	}

	public void setClerk(String clerk) {
		this.clerk = clerk;
	}

	public int getShippriority() {
		return shippriority;
	}

	public void setShippriority(int shippriority) {
		this.shippriority = shippriority;
	}

	public String getComment() {
		return comment;
	}

	public void setComment(String comment) {
		this.comment = comment;
	}

	@Override
	public ArrayList<RootData> generateObjects(int number) {
		ArrayList<RootData> tmpList = new ArrayList<RootData>();

		List<RootData> rootDataLineItems = new ArrayList<RootData>();
		List<LineItem> lineItems = new ArrayList<LineItem>();

		rootDataLineItems = new LineItem().generateObjects(10);
		for (RootData tmp : rootDataLineItems) {
			lineItems.add((LineItem) tmp);
		}

		for (int i = 0; i < number; i++) {
			// Order(int , int , char , int , String , char , char , int ,
			// String )
			String ichar = i + "";
			double totalprice = (double) (i);
			Order myOrder = new Order(lineItems, i, i, ichar, totalprice, "" + i, ichar, ichar, i, "" + i);

			tmpList.add(myOrder);
		}

		return tmpList;
	}

	public Order readJSONSupplier(JsonObject supplierJsonObject) {

		Order m_order = new Order();
		m_order.setOrderkey(supplierJsonObject.getInt("orderkey"));
		m_order.setCustkey(supplierJsonObject.getInt("custkey"));

		m_order.setOrderstatus(supplierJsonObject.getString("orderstatus"));

		m_order.setTotalprice(supplierJsonObject.getInt("totalprice"));
		m_order.setOrderdate(supplierJsonObject.getString("orderdate"));
		m_order.setOrderpriority(supplierJsonObject.getString("orderpriority"));
		m_order.setClerk(supplierJsonObject.getString("clerk"));
		m_order.setShippriority(supplierJsonObject.getInt("shippriority"));
		m_order.setComment(supplierJsonObject.getString("comment"));

		return m_order;
	}

	public List<LineItem> getLineItems() {
		return lineItems;
	}

	public void setLineItems(List<LineItem> lineItems) {
		this.lineItems = lineItems;
	}

	@Override
	public int compareTo(RootData data) {
		Order m_Order = (Order) data;
		int result=1;
		
		if (m_Order.getOrderkey() == this.getOrderkey()) {
			result=0;
			for (int i = 0; i < m_Order.lineItems.size(); i++) {
					result=m_Order.lineItems.get(i).compareTo(this.lineItems.get(i));
			}
		}
		return result;
	}

	@Override
	public byte[] jsonSerialization() {

		// Set the properties
		Map<String, Object> properties = new HashMap<String, Object>(1);
		properties.put(JsonGenerator.PRETTY_PRINTING, true);

		JsonWriterFactory writerFactory = Json.createWriterFactory(properties);

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		try {
			GZIPOutputStream gzipOutputStream = new GZIPOutputStream(bos);
			JsonWriter jsonWriter = writerFactory.createWriter(gzipOutputStream);
			jsonWriter.writeObject(this.jsonObjectBuilder());
			jsonWriter.close();
		} catch (IOException e) {
			logger.error(e);
		}
		// Get the bytes of the serialized JSON object
		byte[] buf = bos.toByteArray();
		return buf;
	}

	public JsonObject jsonObjectBuilder() {
		JsonObjectBuilder orderJsonObjectBuilder = Json.createObjectBuilder();

		// add all of the attributes of LineItem
		orderJsonObjectBuilder.add("orderkey", this.getOrderkey());
		orderJsonObjectBuilder.add("custkey", this.getCustkey());
		orderJsonObjectBuilder.add("orderstatus", this.getOrderstatus());
		orderJsonObjectBuilder.add("totalprice", this.getTotalprice());
		orderJsonObjectBuilder.add("orderdate", this.getOrderdate());
		orderJsonObjectBuilder.add("orderpriority", this.getOrderpriority());
		orderJsonObjectBuilder.add("clerk", this.getClerk());
		orderJsonObjectBuilder.add("shippriority", this.getShippriority());
		orderJsonObjectBuilder.add("comment", this.getComment());

		// Order are a list
		JsonArrayBuilder jsonOrdersArray = Json.createArrayBuilder();

		for (LineItem myLineItem : this.lineItems) {

			jsonOrdersArray.add(myLineItem.jsonObjectBuilder());
		}
		// add the Parts as JSON array to the lineitemBuilder
		orderJsonObjectBuilder.add("lineItems", jsonOrdersArray);

		JsonObject lineitemJsonObject = orderJsonObjectBuilder.build();
		return lineitemJsonObject;
	}

	@Override
	public RootData jsonDeserialization(byte[] buf) {

		Order myOrder = new Order();

		ByteArrayInputStream b = new ByteArrayInputStream(buf);

		try {
			InputStream gzis = new GZIPInputStream(b);
			// create JsonReader object
			JsonReader jsonReader = Json.createReader(gzis);

			// get JsonObject from JsonReader
			JsonObject jsonObject = jsonReader.readObject();
			// we can close IO resource and JsonReader now
			jsonReader.close();
			b.close();

			myOrder = readJSONLineItem(jsonObject);

			// we can close IO resource and JsonReader now
			jsonReader.close();

		} catch (Exception e) {
			logger.error("Can not read JSON object Part from byteArray", e);
		}

		return myOrder;

	}

	public Order readJSONLineItem(JsonObject jsonObject) {
		Order myOrder = new Order();

		myOrder.setOrderkey(jsonObject.getInt("orderkey"));
		myOrder.setCustkey(jsonObject.getInt("custkey"));
		myOrder.setOrderstatus(jsonObject.getString("orderstatus"));
		myOrder.setTotalprice(jsonObject.getInt("totalprice"));
		myOrder.setOrderdate(jsonObject.getString("orderdate"));
		myOrder.setOrderpriority(jsonObject.getString("orderpriority"));
		myOrder.setClerk(jsonObject.getString("clerk"));
		myOrder.setShippriority(jsonObject.getInt("shippriority"));
		myOrder.setComment(jsonObject.getString("comment"));

		// reading LineItem arrays from json
		JsonArray jsonSupplier = jsonObject.getJsonArray("lineItems");
		ArrayList<LineItem> lineItems = new ArrayList<LineItem>();
		LineItem myLineItem = new LineItem();

		for (int i = 0; i < jsonSupplier.size(); i++) {
			JsonObject m_LineItemJsonObject = jsonSupplier.getJsonObject(i);
			lineItems.add(myLineItem.readJSONLineItem(m_LineItemJsonObject));
		}
		myOrder.setLineItems(lineItems);

		return myOrder;
	}

	// TODO - it is not need to implement protocol buffer for this class
	@Override
	public byte[] protocolBufferWrite() {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public RootData protocolBufferRead(byte[] buf) {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public Order javaDefaultDeserialization(byte[] buf) {
		ByteArrayInputStream b = new ByteArrayInputStream(buf);
		ObjectInputStream objectInputStream;

		Order p = null;

		try {
			objectInputStream = new ObjectInputStream(b);
			p = (Order) objectInputStream.readObject();

		} catch (IOException | ClassNotFoundException e) {
			logger.error("Can read object Part from byteArray", e);
		}
		return p;
	}

	@Override
	public byte[] writeByteBuffer() {

		// private List<LineItem> lineItems;
		//
		// private int orderkey; // 4 bytes
		// private int custkey; // 4 bytes
		// private int totalprice; // 4 bytes
		// private int shippriority; //4 bytes
		//
		// private String orderstatus;
		// private String orderdate;
		// private String orderpriority;
		// private String clerk;
		// private String comment;

		// Container
		byte[] containerBytes = orderstatus.getBytes(Charset.forName("UTF-8"));
		byte[] nameBytes = orderdate.getBytes(Charset.forName("UTF-8"));
		byte[] mfgrBytes = orderpriority.getBytes(Charset.forName("UTF-8"));
		byte[] brandBytes = clerk.getBytes(Charset.forName("UTF-8"));
		byte[] commentBytes = comment.getBytes(Charset.forName("UTF-8"));

		ArrayList<byte[]> lineItemBytes = new ArrayList<byte[]>();

		int lengthOfLineItems = 0;

		for (LineItem tmpLineItem : lineItems) {
			byte[] tmpLineItemBytes = tmpLineItem.writeByteBuffer();
			lineItemBytes.add(tmpLineItemBytes);

			// 4 bytes is for an additional integer for write the size byteArray
			lengthOfLineItems = lengthOfLineItems + tmpLineItemBytes.length + 4;
		}

		// allocate a buffer to write this object
		// an additional 4 bytes for the number of LineItem in the list
		ByteBuffer byteBuffer = ByteBuffer.allocate(44 + containerBytes.length + nameBytes.length + mfgrBytes.length + brandBytes.length + commentBytes.length + lengthOfLineItems);

		byteBuffer.putInt(orderkey); // 4 bytes
		byteBuffer.putInt(custkey); // 4 bytes
		byteBuffer.putDouble(totalprice); // 8 bytes
		byteBuffer.putInt(shippriority); // 4 bytes

		byteBuffer.putInt(containerBytes.length); // 4 bytes size of
		byteBuffer.put(containerBytes);

		// name
		byteBuffer.putInt(nameBytes.length); // 4 bytes size of
		byteBuffer.put(nameBytes);

		// mfgr
		byteBuffer.putInt(mfgrBytes.length); // 4 bytes size of
		byteBuffer.put(mfgrBytes);

		// brand
		byteBuffer.putInt(brandBytes.length); // 4 bytes size of
		byteBuffer.put(brandBytes);

		// comment
		byteBuffer.putInt(commentBytes.length); // 4 bytes size of
		byteBuffer.put(commentBytes);

		// first write how many items have to be read from the buffer
		// number of LineItems in the list
		byteBuffer.putInt(lineItemBytes.size());

		for (byte[] mylineItemBytes : lineItemBytes) {
			// Add each lineItem byte array one bye one
			byteBuffer.putInt(mylineItemBytes.length);
			byteBuffer.put(mylineItemBytes);
		}

		return byteBuffer.array();
	}

	@Override
	public RootData readByteBuffer(byte[] buf) {

		// private List<LineItem> lineItems;
		//
		// private int orderkey; // 4 bytes
		// private int custkey; // 4 bytes
		// private int totalprice; // 4 bytes
		// private int shippriority; //4 bytes
		//
		// private String orderstatus;
		// private String orderdate;
		// private String orderpriority;
		// private String clerk;
		// private String comment;

		// allocate a buffer for 12 bytes to write an integer and a double
		ByteBuffer byteBuffer = ByteBuffer.wrap(buf);

		int tmpOrderkey = byteBuffer.getInt(); // 4 bytes
		int tmpCustkey = byteBuffer.getInt(); // 4 bytes
		double tmpTotalprice = byteBuffer.getDouble(); // 8 bytes
		int tmpShippriority = byteBuffer.getInt(); // 4 bytes
		int stringSize;

		// Get orderstatus
		stringSize = byteBuffer.getInt(); // 4 bytes = 20
		String tmpOrderstatus = extractString(byteBuffer, stringSize);

		// Get orderdate
		stringSize = byteBuffer.getInt(); // 4 bytes
		String tmpOrderdate = extractString(byteBuffer, stringSize);

		// orderpriority
		stringSize = byteBuffer.getInt(); // 4 bytes = 24
		String tmpOrderpriority = extractString(byteBuffer, stringSize);

		// clerk
		stringSize = byteBuffer.getInt(); // 4 bytes = 24
		String tmpClerk = extractString(byteBuffer, stringSize);

		// comment
		stringSize = byteBuffer.getInt(); // 4 bytes = 24
		String tmpComment = extractString(byteBuffer, stringSize);

		int numberOfLineItems = byteBuffer.getInt();
		ArrayList<LineItem> myListOfLineItems = new ArrayList<LineItem>();

		for (int i = 0; i < numberOfLineItems; i++) {

			// add LineItems one by one
			byte[] LineItemBytes = new byte[byteBuffer.getInt()];
			byteBuffer.get(LineItemBytes, 0, LineItemBytes.length);
			LineItem tmpLineItem = (LineItem) new LineItem().readByteBuffer(LineItemBytes);
			myListOfLineItems.add(tmpLineItem);
		}

		return new Order(myListOfLineItems, tmpOrderkey, tmpCustkey, tmpOrderstatus, tmpTotalprice, tmpOrderdate, tmpOrderpriority, tmpClerk, tmpShippriority, tmpComment);

	}

}

File: C++/Order.cpp
#include<iostream>
#include<cstring>
#include <sstream>
#include <vector>
#include "Order.h"
#include "LineItem.h"
using namespace std;


	// CREATE TABLE ORDERS ( O_ORDERKEY INTEGER NOT NULL,
	// O_CUSTKEY INTEGER NOT NULL,
	// O_ORDERSTATUS CHAR(1) NOT NULL,
	// O_TOTALPRICE DECIMAL(15,2) NOT NULL,
	// O_ORDERDATE DATE NOT NULL,
	// O_ORDERPRIORITY CHAR(15) NOT NULL,
	// O_CLERK CHAR(15) NOT NULL,
	// O_SHIPPRIORITY INTEGER NOT NULL,
	// O_COMMENT VARCHAR(79) NOT NULL);



	//Default constructor:
	Order::Order() {	
	}


	//Constructor with arguments:
	Order::Order(vector<LineItem*> lineItems, int orderKey, int custKey, string orderStatus, double totalPrice, string orderDate, string orderPriority, string clerk, int shipPriority, string comment) {
		this->lineItems = lineItems;
		this->orderKey = orderKey;
		this->custKey = custKey;
		this->orderStatus = orderStatus;
		this->totalPrice = totalPrice;
		this->orderDate = orderDate;
		this->orderPriority = orderPriority;
		this->clerk = clerk;
		this->shipPriority = shipPriority;
		this->comment = comment;
	}


	//Getters and Setters Follow:
	vector<LineItem*> Order::getLineItems(){
		return lineItems;
	}

	void Order::setLineItems(vector<LineItem*> lineItems) {
		this->lineItems = lineItems;
	}

	int Order::getOrderKey() {
		return orderKey;
	}

	void Order::setOrderKey(int orderKey) {
		this->orderKey = orderKey;
	}

	int Order::getCustKey() {
		return custKey;
	}

	void Order::setCustKey(int custKey) {
		this->custKey = custKey;
	}

	double Order::getTotalPrice() {
		return totalPrice;
	}

	void Order::setTotalPrice(double totalPrice) {
		this->totalPrice = totalPrice;
	}


	string Order::getOrderStatus() {
		return orderStatus;
	}

	 void Order::setOrderStatus(string orderStatus) {
		this->orderStatus = orderStatus;
	}

	string Order::getOrderDate() {
		return orderDate;
	}

	void Order::setOrderDate(string orderDate) {
		this->orderDate = orderDate;
	}

	string Order::getOrderPriority() {
		return orderPriority;
	}

	void Order::setOrderPriority(string orderPriority) {
		this->orderPriority = orderPriority;
	}

	string Order::getClerk() {
		return clerk;
	}

	void Order::setClerk(string clerk) {
		this->clerk = clerk;
	}

	int Order::getShipPriority() {
		return shipPriority;
	}

	void Order::setShipPriority(int shipPriority) {
		this->shipPriority = shipPriority;
	}

	string Order::getComment() {
		return comment;
	}

	void Order::setComment(string comment) {
		this->comment = comment;
	}			



	//C++: Explicit call needed for printing content:
	string Order::toString() {
		string stringS = itos(this->getOrderKey()) + " - " + itos(this->getCustKey())  + " - " + this->getOrderStatus() + " - " + dtos(this->getTotalPrice()) + " - " + this->getOrderDate()+ " - " + this->getOrderPriority()
		 + " - " + this->getClerk() + " - " + itos(this->getShipPriority()) + " - " + this->getComment() ; 
		;
	
		//All lineitems:
		for (int i = 0; i < lineItems.size(); i++) {
			stringS += " LineItem " + itos(i) + " : " + lineItems.at(i)->toString();
		}

		return stringS;
	}





	//Use for generating objects in bulk:
	vector<Order *> Order::generateObjects(int number) {
		//Limiting the number of line items.
		int numLineItems = 4;
		vector<Order *> orderList;
		for (int j = 0; j < number; j++) {
			vector<LineItem *> lineItemList = LineItem::generateObjects(2 + (j%numLineItems));
			Order * t = new Order(lineItemList, j, j+1, "OrderStatus-" + itos(j+2),  j+3.0, "OrderDate-" + itos(j+4), "OrderPriority-" + itos(j+5), "Clerk-" + itos(j+6), j+7, "comment-" + itos(j+8));
			orderList.push_back(t);
		}
		return orderList;
	}


	/*
	 * Functions related to Hand Coded Serialization.
	 */
	char * Order::serialize(char * buffer, int& objectSize) {

		//First determine the size of object:
		//Integer elements:
		objectSize = 3 * sizeof(int);
		//Double elements:
		objectSize += 1 * sizeof(double);
		//String sizes: 
		objectSize += 5 * sizeof(int);
		//String Elements:
		objectSize += (orderStatus.length() + orderDate.length() + orderPriority.length() + clerk.length() + comment.length());		
		
		//Vector of lineitems:
		objectSize += sizeof(int);

		//Now, serialize the object.
		char * tempBuffer = buffer;
		
		//Copy Integers:
		tempBuffer = copyInt(tempBuffer, orderKey);	
		tempBuffer = copyInt(tempBuffer, custKey);	
		tempBuffer = copyInt(tempBuffer, shipPriority);

		//Copy Doubles:		
		tempBuffer = copyDouble(tempBuffer, totalPrice);

		//Copy Strings:
		tempBuffer = copyString(tempBuffer, orderStatus);		
		tempBuffer = copyString(tempBuffer, orderDate);
		tempBuffer = copyString(tempBuffer, orderPriority);
		tempBuffer = copyString(tempBuffer, clerk);
		tempBuffer = copyString(tempBuffer, comment);


		//Copy encapsulated objects:
		int numLineItems =  lineItems.size();
		tempBuffer = copyInt(tempBuffer, numLineItems);	


		//New code:
		int lineItemSize;
		for (int i = 0; i < lineItems.size(); i++) {
			lineItemSize = 0;
			lineItems.at(i)->serialize(tempBuffer, lineItemSize);
			tempBuffer += lineItemSize;
			objectSize += lineItemSize;
		}

		return buffer;
	}




	int Order::deserialize(char * buffer) {
		int bytesRead = 0;
		
		//Integers:
		orderKey = parseInt(buffer+bytesRead); bytesRead += sizeof(orderKey);	
		custKey = parseInt(buffer+bytesRead); bytesRead += sizeof(custKey);	
		shipPriority = parseInt(buffer+bytesRead); bytesRead += sizeof(shipPriority);		

		//Doubles:
		totalPrice = parseDouble(buffer+bytesRead); bytesRead += sizeof(totalPrice);

		//Strings:
		parseString(buffer+bytesRead, orderStatus); bytesRead += (sizeof(int) + orderStatus.length());
		parseString(buffer+bytesRead, orderDate); bytesRead += (sizeof(int) + orderDate.length());
		parseString(buffer+bytesRead, orderPriority); bytesRead += (sizeof(int) + orderPriority.length());
		parseString(buffer+bytesRead, clerk); bytesRead += (sizeof(int) + clerk.length());
		parseString(buffer+bytesRead, comment); bytesRead += (sizeof(int) + comment.length());


		//Encapsulated objects:
		int numLineItems = parseInt(buffer+bytesRead); bytesRead += sizeof(numLineItems);	

		LineItem * tempLineItem;
		for(int i = 0; i < numLineItems; i++) {
			tempLineItem = new LineItem();
			bytesRead += tempLineItem->deserialize(buffer+bytesRead);
			lineItems.push_back(tempLineItem);
		}

		return  bytesRead;
	}	




	//Implement your own custom comparator:
	bool Order::operator< (Order& other) {

		if (orderKey == other.orderKey) {
			bool result = false;

			if (lineItems.size() == other.getLineItems().size()) {
				for (int i = 0; i < lineItems.size(); i++) {
					result = lineItems.at(i) < (other.getLineItems().at(i));
					if (result) return true;
				}
			}

			return lineItems.size() < other.getLineItems().size();		
		}

		return orderKey < other.orderKey;	
	}



	//Create a inplace object:
	OrderIP * Order::createInPlaceObject() {
		return new OrderIP(lineItems, lineItems.size(), orderKey, custKey, orderStatus, totalPrice, orderDate, orderPriority, clerk, shipPriority, comment);			
	}

	//In place serialization:
	long Order::serializeInPlace(char * fileStart) {
		//In place object:
		this->inplace = this->createInPlaceObject();	

		//Return pointer:
		return ((long)inplace - (long)fileStart);
	}